### ISSUE TITLE
Unable to download some mods through Steam Linux version

### ISSUE THREAD

**Issue by xeals** [tModLoader/tModLoader#908]
### Description

Attempting to download most new mods or update certain mods through the Steam version on a Linux system fails due to an SSL error. I am using the Steam runtime so I believe the host system configuration should not be an influence in this. I haven't been able to find any other mention of this exact issue.

Mods I am not able to update:

- Thorium
- JoostMod
- GRealm

I have been unable to download most of the popular mods in the mod browser. I'm unsure what the difference between ones that worked and ones that failed are.

### Log file

https://gist.github.com/xeals/ef04bbdc094990a81f5de5aa8e632ab5

### Steps to reproduce

1. Install Steam version on Linux
2. Launch using Steam runtime
3. Attempt to install or update one of the mods above

### Expected behavior 

Mods should download.

### Actual behavior

Download progress bar appears and after a few seconds at 0%, closes back to the mod listing. Mod is not downloaded.

### Reproduction frequency

100%.

**Comment by PoroCYon** [comment]
Feels like a certificate validation error to me, but both the `github.com` and the S3 (to which the github.com link redirects) certs seem valid to me on first sight (and my browser agrees).

Are you sure the date and time in your system is set up correctly, and that you have all required certificates installed? (eg. the `ca-cacert`, `ca-certificates`, `ca-certificates-mozilla`, `ca-certificates-mono`, `ca-certificates-java` packages should all be installed)

**Comment by xeals** [comment]
> Feels like a certificate validation error to me, but both the `github.com` and the S3 (to which the github.com link redirects) certs seem valid to me on first sight (and my browser agrees).
> 
> Can you try opening the download link in the logfile with your browser, and see if that works?

The links open and download fine in a browser.

> Are you sure the date and time in your system is set up correctly, and that you have all required certificates installed? (eg. the `ca-cacert`, `ca-certificates`, `ca-certificates-mozilla`, `ca-certificates-mono`, `ca-certificates-java` packages should all be installed, if those exist for your distro. Ubuntu and Debian have all of them except the `-mozilla` one, Arch only has a few, Void merges these into just one package.)

My system clock is a few minutes ahead of UTC, but it's never affected anything. And wouldn't that prevent downloading any mods rather than just a few?

```
$ pacman -Ss ca-certificates
core/ca-certificates 20181109-3 [installed]
core/ca-certificates-mozilla 3.52-1 [installed: 3.51.1-1]
core/ca-certificates-utils 20181109-3 [installed]
```

> You can run sudo apt install mono-complete if you want, this will fix these errors. We're looking into a normal fix

```
$ pacman -Qs mono
mono
mono-addins
mono-tools
```

I've already seen those issue reports and tried those fixes, which haven't worked -- that's why I've tried to note that in the OP.

**Comment by AKDRER** [comment]
I am having the same issue, and with more than just those mods. All of them are on github, any other mods download.

**Comment by AKDRER** [comment]
Update-
Spontaneously began working again, pulling instead from dropbox repositories. This happened a few minutes after I had some failed attempts from github, and I never closed the game. I do not know what happened.

**Comment by hagabaka** [comment]
Running tModLoader using the system mono fixes this problem for me. I did this by setting the Launch Options to `%command%-mono` (with no space before the dash) so it runs the script `tModLoader-mono` instead of the `tModLoader`.

**Comment by TheCatster** [comment]
The tModLoader-mono helped me as well. That should be in documentation somewhere. This occurred on Arch and on MacOS for me.

**Comment by tzhouhc** [comment]
Issue persists in the Linux Server (0.11.7.5), copying tModLoader-mono into tModLoaderServer-mono doesn't work even though the server does seem to run and can download some other mods fine (like RecipeBrowser for some reason).

**Comment by pvanheus** [comment]
> Running tModLoader using the system mono fixes this problem for me. I did this by setting the Launch Options to `%command%-mono` (with no space before the dash) so it runs the script `tModLoader-mono` instead of `tModLoader`.

Thanks, this worked for me too (Ubuntu 20.04, installed `mono-complete` package and do this change). I did not know what "Launch Options" were though - for others who have this problem, these are found in the settings for the game on Steam.

**Comment by hauntingEcho** [comment]
I'm seeing this issue as well.  Launching tModLoader from console gives me an output which points to a certificate error.  I'm running Debian Stable.

```
[21:01:25] [1/DEBUG] [tML]: Web Request: http://javid.ddns.net/tModLoader/download.php?Down=mods/BossChecklist.tmod
[21:01:25] [Threadpool worker/DEBUG] [tML]: Web Request: https://www.dropbox.com/s/92umbq5mwzndfql/BossChecklist.tmod?dl=1
[21:01:25] [Threadpool worker/DEBUG] [tML]: Web Request: https://www.dropbox.com/s/dl/92umbq5mwzndfql/BossChecklist.tmod
[21:01:25] [Threadpool worker/DEBUG] [tML]: Web Request: https://ucfbd69811c068f0b584124d40c4.dl.dropboxusercontent.com/cd/0/get/BBdvUIezsCV7BbtmA8YvZ7zajo9yHQ4V__ePEg2ThTIVU1uL6Ne9lyNehk1RQVoBBHHAIcUvEiNYjnjUvixTabi98s3l1YbBNiVZhK5-MmRkma2jVKo6pKdfmRB-D4QGZTs/file?dl=1#
[21:01:25] [Threadpool worker/ERROR] [tML]: There was a problem downloading the mod Boss Checklist
System.AggregateException: One or more errors occurred. ---> System.AggregateException: One or more errors occurred. ---> System.Net.WebException: Error: TrustFailure (The authentication or decryption has failed.) ---> System.IO.IOException: The authentication or decryption has failed. ---> System.IO.IOException: The authentication or decryption has failed. ---> Mono.Security.Protocol.Tls.TlsException: Invalid certificate received from server. Error code: 0xffffffff800b010a
  at Mono.Security.Protocol.Tls.RecordProtocol.EndReceiveRecord (IAsyncResult asyncResult) [0x00040] in Mono.Security.dll 
  at Mono.Security.Protocol.Tls.SslClientStream.SafeEndReceiveRecord (IAsyncResult ar, Boolean ignoreEmpty) [0x00000] in Mono.Security.dll 
  at Mono.Security.Protocol.Tls.SslClientStream.NegotiateAsyncWorker (IAsyncResult result) [0x00071] in Mono.Security.dll 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslClientStream.EndNegotiateHandshake (IAsyncResult result) [0x0003b] in Mono.Security.dll 
  at Mono.Security.Protocol.Tls.SslStreamBase.AsyncHandshakeCallback (IAsyncResult asyncResult) [0x0000c] in Mono.Security.dll 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslStreamBase.EndRead (IAsyncResult asyncResult) [0x00057] in Mono.Security.dll 
  at Mono.Net.Security.Private.LegacySslStream.EndAuthenticateAsClient (IAsyncResult asyncResult) [0x00011] in System.dll 
  at Mono.Net.Security.Private.LegacySslStream.AuthenticateAsClient (String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation) [0x0000e] in System.dll 
  at Mono.Net.Security.MonoTlsStream.CreateStream (Byte[] buffer) [0x00044] in System.dll 
   --- End of inner exception stack trace ---
  at System.Net.HttpWebRequest.EndGetResponse (IAsyncResult asyncResult) [0x00064] in System.dll 
  at Terraria.ModLoader.UI.DownloadManager.DownloadFile+<>c__DisplayClass21_0.<Download>b__0 (IAsyncResult asyncResult) [0x0000b] in tModLoader.exe 
  at System.Threading.Tasks.TaskFactory[TResult].FromAsyncCoreLogic (IAsyncResult iar, Func[T,TResult] endFunction, Action[T] endAction, Task[TResult] promise, Boolean requiresSynchronization) [0x00014] in mscorlib.dll 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (Boolean includeTaskCanceledExceptions) [0x00014] in mscorlib.dll 
  at System.Threading.Tasks.Task[TResult].GetResultCore (Boolean waitCompletionNotification) [0x00034] in mscorlib.dll 
  at System.Threading.Tasks.Task[TResult].get_Result () [0x0000b] in mscorlib.dll 
  at Terraria.ModLoader.UI.DownloadManager.DownloadFile+<>c__DisplayClass21_0.<Download>b__1 (Task[TResult] t) [0x00000] in tModLoader.exe 
  at System.Threading.Tasks.ContinuationResultTaskFromResultTask[TAntecedentResult,TResult].InnerInvoke () [0x00027] in mscorlib.dll 
  at System.Threading.Tasks.Task.Execute () [0x00016] in mscorlib.dll 
   --- End of inner exception stack trace ---

---> (Inner Exception #0) System.AggregateException: One or more errors occurred. ---> System.Net.WebException: Error: TrustFailure (The authentication or decryption has failed.) ---> System.IO.IOException: The authentication or decryption has failed. ---> System.IO.IOException: The authentication or decryption has failed. ---> Mono.Security.Protocol.Tls.TlsException: Invalid certificate received from server. Error code: 0xffffffff800b010a
  at Mono.Security.Protocol.Tls.RecordProtocol.EndReceiveRecord (IAsyncResult asyncResult) [0x00040] in Mono.Security.dll 
  at Mono.Security.Protocol.Tls.SslClientStream.SafeEndReceiveRecord (IAsyncResult ar, Boolean ignoreEmpty) [0x00000] in Mono.Security.dll 
  at Mono.Security.Protocol.Tls.SslClientStream.NegotiateAsyncWorker (IAsyncResult result) [0x00071] in Mono.Security.dll 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslClientStream.EndNegotiateHandshake (IAsyncResult result) [0x0003b] in Mono.Security.dll 
  at Mono.Security.Protocol.Tls.SslStreamBase.AsyncHandshakeCallback (IAsyncResult asyncResult) [0x0000c] in Mono.Security.dll 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslStreamBase.EndRead (IAsyncResult asyncResult) [0x00057] in Mono.Security.dll 
  at Mono.Net.Security.Private.LegacySslStream.EndAuthenticateAsClient (IAsyncResult asyncResult) [0x00011] in System.dll 
  at Mono.Net.Security.Private.LegacySslStream.AuthenticateAsClient (String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation) [0x0000e] in System.dll 
  at Mono.Net.Security.MonoTlsStream.CreateStream (Byte[] buffer) [0x00044] in System.dll 
   --- End of inner exception stack trace ---
  at System.Net.HttpWebRequest.EndGetResponse (IAsyncResult asyncResult) [0x00064] in System.dll 
  at Terraria.ModLoader.UI.DownloadManager.DownloadFile+<>c__DisplayClass21_0.<Download>b__0 (IAsyncResult asyncResult) [0x0000b] in tModLoader.exe 
  at System.Threading.Tasks.TaskFactory[TResult].FromAsyncCoreLogic (IAsyncResult iar, Func[T,TResult] endFunction, Action[T] endAction, Task[TResult] promise, Boolean requiresSynchronization) [0x00014] in mscorlib.dll 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (Boolean includeTaskCanceledExceptions) [0x00014] in mscorlib.dll 
  at System.Threading.Tasks.Task[TResult].GetResultCore (Boolean waitCompletionNotification) [0x00034] in mscorlib.dll 
  at System.Threading.Tasks.Task[TResult].get_Result () [0x0000b] in mscorlib.dll 
  at Terraria.ModLoader.UI.DownloadManager.DownloadFile+<>c__DisplayClass21_0.<Download>b__1 (Task[TResult] t) [0x00000] in tModLoader.exe 
  at System.Threading.Tasks.ContinuationResultTaskFromResultTask[TAntecedentResult,TResult].InnerInvoke () [0x00027] in mscorlib.dll 
  at System.Threading.Tasks.Task.Execute () [0x00016] in mscorlib.dll 
---> (Inner Exception #0) System.Net.WebException: Error: TrustFailure (The authentication or decryption has failed.) ---> System.IO.IOException: The authentication or decryption has failed. ---> System.IO.IOException: The authentication or decryption has failed. ---> Mono.Security.Protocol.Tls.TlsException: Invalid certificate received from server. Error code: 0xffffffff800b010a
  at Mono.Security.Protocol.Tls.RecordProtocol.EndReceiveRecord (IAsyncResult asyncResult) [0x00040] in Mono.Security.dll 
  at Mono.Security.Protocol.Tls.SslClientStream.SafeEndReceiveRecord (IAsyncResult ar, Boolean ignoreEmpty) [0x00000] in Mono.Security.dll 
  at Mono.Security.Protocol.Tls.SslClientStream.NegotiateAsyncWorker (IAsyncResult result) [0x00071] in Mono.Security.dll 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslClientStream.EndNegotiateHandshake (IAsyncResult result) [0x0003b] in Mono.Security.dll 
  at Mono.Security.Protocol.Tls.SslStreamBase.AsyncHandshakeCallback (IAsyncResult asyncResult) [0x0000c] in Mono.Security.dll 
   --- End of inner exception stack trace ---
  at Mono.Security.Protocol.Tls.SslStreamBase.EndRead (IAsyncResult asyncResult) [0x00057] in Mono.Security.dll 
  at Mono.Net.Security.Private.LegacySslStream.EndAuthenticateAsClient (IAsyncResult asyncResult) [0x00011] in System.dll 
  at Mono.Net.Security.Private.LegacySslStream.AuthenticateAsClient (String targetHost, X509CertificateCollection clientCertificates, SslProtocols enabledSslProtocols, Boolean checkCertificateRevocation) [0x0000e] in System.dll 
  at Mono.Net.Security.MonoTlsStream.CreateStream (Byte[] buffer) [0x00044] in System.dll 
   --- End of inner exception stack trace ---
  at System.Net.HttpWebRequest.EndGetResponse (IAsyncResult asyncResult) [0x00064] in System.dll 
  at Terraria.ModLoader.UI.DownloadManager.DownloadFile+<>c__DisplayClass21_0.<Download>b__0 (IAsyncResult asyncResult) [0x0000b] in tModLoader.exe 
  at System.Threading.Tasks.TaskFactory[TResult].FromAsyncCoreLogic (IAsyncResult iar, Func[T,TResult] endFunction, Action[T] endAction, Task[TResult] promise, Boolean requiresSynchronization) [0x00014] in mscorlib.dll <---
<---
```

**Comment by hauntingEcho** [comment]
Oddly enough, after having installed `mono-complete` and then installed some modes from `tModLoader-mono`, `tModLoader-kick` now behaves as expected.

**Comment by Redd56** [comment]
ive had an issue with almost all mods not downloading

**Comment by cSDes1gn** [comment]
> Oddly enough, after having installed `mono-complete` and then installed some mods from `tModLoader-mono`, `tModLoader-kick` now behaves as expected.

+1 to this

**Comment by HellishINC** [comment]
`%command%-mono` also fixes this for me running Arch.

**Comment by Mirsario** [comment]
This issue should now be fixed within 1.4 TML, as Mono has been replaced with .NET 5, and mod browser code is now quite different with the addition of steam workshop support. An alpha version of 1.4 TML can be found in [our discord](https://discord.gg/RMZCqq6).

Meanwhile, for 1.3:
> Use system Mono, or manual downloads from http://javid.ddns.net/tModLoader/DirectModDownloadListing.php

**Comment by HellishINC** [comment]
Is there any way builds could be hosted outside of discord?
